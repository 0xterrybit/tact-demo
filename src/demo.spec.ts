import { Address, OpenedContract, toNano } from "@ton/core";
import { ContractSystem, Treasure } from "@tact-lang/emulator";
import { did } from "./output/rns_did";

describe("did contract", () => {

  let system: ContractSystem;
  let contract: OpenedContract<did>;

  let owner: Treasure;
  let nonOwner: Treasure;

  beforeEach(async () => {
    
    system = await ContractSystem.create();
    owner = system.treasure("owner");
    nonOwner = system.treasure("non-owner");

    const ins = await did.fromInit(owner.address);
    contract = system.open(ins);
    system.name(contract.address, "main");


  });

  it("should deploy correctly", async () => {
    const track = system.track(contract);
    await contract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });
    await system.run();
    expect(track.collect()).toMatchSnapshot();
  });

  it("should increment counter by owner", async () => {

    const track = system.track(contract);
    await contract.send(owner, { value: toNano(1) }, "increment");
    await system.run();

    expect(track.collect()).toMatchSnapshot();
    expect(await contract.getCounter()).toEqual(1n);

  });

  it("should not increment counter by non-owner", async () => {

    const track = system.track(contract);

    await contract.send(nonOwner, { value: toNano(1) }, "increment");
    await system.run();

    expect(track.collect()).toMatchSnapshot();
    expect(await contract.getCounter()).toEqual(0n);

  });
});
