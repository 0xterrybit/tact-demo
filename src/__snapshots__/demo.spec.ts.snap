// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`did contract should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "type": "known",
            "value": {
              "$$type": "Deploy",
              "queryId": 0n,
            },
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@main",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8040n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "DeployOk",
                "queryId": 0n,
              },
            },
            "bounce": false,
            "from": "@main",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": "0.990764",
          },
        ],
      },
    ],
  },
]
`;

exports[`did contract should increment counter by owner 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "increment",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@main",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8411n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "incremented",
              "type": "text",
            },
            "bounce": true,
            "from": "@main",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": "0.990369",
          },
        ],
      },
    ],
  },
]
`;

exports[`did contract should not increment counter by non-owner 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "increment",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(non-owner)",
          "to": "@main",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "failed",
        "errorCode": 4429,
        "errorMessage": "Invalid sender",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF00000000696E6372656D656E74}",
            "type": "cell",
          },
          "bounce": false,
          "from": "@main",
          "to": "@treasure(non-owner)",
          "type": "internal",
          "value": "0.994877",
        },
      },
    ],
  },
]
`;
